airbyte-integrations/bases/source-test-lib/src/main/java/io/airbyte/integrations/base/TestSource.java

```diff
@@ -0,0 +1,265 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public abstract class TestSource {
+
+  private TestDestinationEnv testEnv;
+
+  private Path jobRoot;
+  protected Path localRoot;
+  private ProcessBuilderFactory pbf;
+
+  /**
+   * Name of the docker image that the tests will run against.
+   *
+   * @return docker image name
+   */
+  protected abstract String getImageName();
+
+  /**
+   * Configuration specific to the integration. Will be passed to integration where appropriate in
+   * each test. Should be valid.
+   *
+   * @return integration-specific configuration
+   */
+  protected abstract JsonNode getConfig() throws Exception;
+
+  /**
+   * Catalog to be used when attempting read operations.
+   *
+   * @return the catalog
+   * @throws Exception - do what must be done.
+   */
+  protected abstract AirbyteCatalog getCatalog() throws Exception;
+
+  /**
+   * Function that performs any setup of external resources required for the test. e.g. instantiate a
+   * postgres database. This function will be called before EACH test.
+   *
+   * @param testEnv - information about the test environment.
+   * @throws Exception - can throw any exception, test framework will handle.
+   */
+  protected abstract void setup(TestDestinationEnv testEnv) throws Exception;
+
+  /**
+   * Function that performs any clean up of external resources required for the test. e.g. delete a
+   * postgres database. This function will be called after EACH test. It MUST remove all data in the
+   * destination so that there is no contamination across tests.
+   *
+   * @param testEnv - information about the test environment.
+   * @throws Exception - can throw any exception, test framework will handle.
+   */
+  protected abstract void tearDown(TestDestinationEnv testEnv) throws Exception;
+
+  @Before
+  public void setUpInternal() throws Exception {
+    Path testDir = Path.of("/tmp/airbyte_tests/");
+    Files.createDirectories(testDir);
+    final Path workspaceRoot = Files.createTempDirectory(testDir, "test");
+    jobRoot = Files.createDirectories(Path.of(workspaceRoot.toString(), "job"));
+    localRoot = Files.createTempDirectory(testDir, "output");
+    testEnv = new TestDestinationEnv(localRoot);
+
+    setup(testEnv);
+
+    pbf = new DockerProcessBuilderFactory(
+        workspaceRoot,
+        workspaceRoot.toString(),
+        localRoot.toString(),
+        "host");
+  }
+
+  @After
+  public void tearDownInternal() throws Exception {
+    tearDown(testEnv);
+  }
+
+  /**
+   * Verify that when the integrations returns a valid spec.
+   */
+  @Test
+  public void testGetSpec() {
+    final OutputAndStatus<StandardGetSpecOutput> output = runSpec();
+    assertTrue(output.getOutput().isPresent());
```

Comment:
why not also assert that the spec is what we expect?

Reply:
same as for check and discover. i'll add this comment ehre as well

```
    // the worker validates that it is a valid catalog, so we do not need to validate again (as long as
    // we use the worker, which we will not want to do long term).
```

Reply:
we could. i was trying to decrease how much stuff had to be passed into this thing. but i think that was mostly because i was scared about figuring out how that was going to be done. now that it's figured out, it's not too bad. yeah. i think you're right, i'll add it.